cli/flox-rust-sdk/src/providers/flox_cpp_utils.rs

```diff
@@ -0,0 +1,281 @@
+}
+
+/// Rust representation of the output of `pkgdb lock-flake-installable`
+/// This is a direct translation of the definition in
+/// `<flox>/pkgdb/include/flox/lock-flake-installable.hh`
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+#[cfg_attr(test, derive(proptest_derive::Arbitrary))]
+pub struct LockedInstallable {
+    /// locked url of the flakeref component of the installable
+    pub(crate) locked_url: String,
+    pub(crate) flake_description: Option<String>,
+    pub(crate) locked_flake_attr_path: String,
+    pub(crate) derivation: String,
+    /// Map of output names to their paths
+    /// The values are expected to be nix store paths
+    pub(crate) outputs: BTreeMap<String, String>,
+    /// List of output names in the original order
+    pub(crate) output_names: Vec<String>,
+    /// List of output names to install as defined by the package
+    pub(crate) outputs_to_install: Option<Vec<String>>,
+    /// List of output names to install as requested by the user
+    pub(crate) requested_outputs_to_install: Option<Vec<String>>,
+    /// System as defined by the package
+    pub(crate) package_system: String,
+    /// System as defined by the specified target system
+    pub(crate) locked_system: String,
+    pub(crate) name: String,
+    pub(crate) pname: Option<String>,
+    pub(crate) version: Option<String>,
+    pub(crate) description: Option<String>,
+    pub(crate) licenses: Option<Vec<String>>,
+    pub(crate) broken: Option<bool>,
+    pub(crate) unfree: Option<bool>,
+}
+
+/// Required functionality to lock a flake installable
+///
+/// Implemented as a trait to allow mocking as evaluation is
+/// time-consuming unless cached.
+///
+/// The trait is implemented by the [`Pkgdb`] struct which is the canonical implementation
+/// using the `pkgdb lock-flake-installable` command.
+///
+/// The trait is also implemented by the [`LockFlakeInstallableMock`] struct which is used for testing.
+#[enum_dispatch]
+pub trait LockFlakeInstallableTrait {
+    fn lock_flake_installable(
+        &self,
+        system: impl AsRef<str>,
+        installable: impl AsRef<str>,
+    ) -> Result<LockedInstallable, FlakeInstallableError>;
+}
+
+#[derive(Debug)]
+#[enum_dispatch(LockFlakeInstallableTrait)]
+pub enum LockFlakeInstallable {
+    Pkgdb(Pkgdb),
+    Mock(LockFlakeInstallableMock),
+}
+
+impl Default for LockFlakeInstallable {
+    fn default() -> Self {
+        LockFlakeInstallable::Pkgdb(Pkgdb::default())
+    }
+}
+/// A wrapper for (eventually) various `pkgdb` commands
+/// Currently only implements [LockFlakeInstallableTrait] through
+/// `pkgdb lock-flake-installable`.
+#[derive(Debug)]
+pub struct Pkgdb {
+    pkgdb_path: PathBuf,
+}
+
+impl Pkgdb {
+    fn new() -> Self {
+        Pkgdb {
+            pkgdb_path: PathBuf::from(&*PKGDB_BIN),
+        }
+    }
+}
+
+impl Default for Pkgdb {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
```

Comment:
nonblocking: is there any way you'd foresee us not using the default `pkgdb_path`? Then can we just drop it?
```suggestion
```

Reply:
yeah we can make the `Pkgdb struct a union type without any field as well.
Im a bit allergic to globals but I also cant argue with any foreseeable case that we'd need this to be configurable beyond env vars.